### ISSUE TITLE
cocos2d::ui::ScrollView stop scrolling onExit()

### ISSUE THREAD

**Issue by AIGRIND-LLC** [cocos2d/cocos2d-x#18551]
Hello,
Is some cases `ScrollView` may leave the scene while scrolling. For example the user has moved `ScrollView` making it to auto-scroll and closed the whole window that contains this `ScrollView`. Next time the window appears the scene this `ScrollView` will be stuck in scrolling state because it has not stopped to scroll before leaving the scene. Even scroll bars are visible in that state.

Also there is another problem with cocos2d::ui::ScrollView::stopAutoScroll() method.
It does not call corresponding scroll event the game code expects to receive (for example `cocos2d.ui.ScrollView.EventType.AUTOSCROLL_ENDED` event).

Moreover, why is only method to stop auto scrolling provided? In some cases it is required to stop general scrolling.

**Comment by AIGRIND-LLC** [comment]
I've also emulated the situation when `ScrollView` is stuck in scrolling mode,
So you may easily reproduce it.
You need to modify `bool UIScrollViewTest_Both::init()` in **cpp-test** project in this way (only code after comment _"Adding keyboard listener to reproduce scroll stopping issue!"_ is added):
```cpp
bool UIScrollViewTest_Both::init()
{
    if (UIScene::init())
    {
        Size widgetSize = _widget->getContentSize();
        
        // Add a label in which the dragpanel events will be displayed
        _displayValueLabel = Text::create("Move by any direction","fonts/Marker Felt.ttf",32);
        _displayValueLabel->setAnchorPoint(Vec2(0.5f, -1.0f));
        _displayValueLabel->setPosition(Vec2(widgetSize.width / 2.0f, widgetSize.height / 2.0f + _displayValueLabel->getContentSize().height * 1.5f));
        _uiLayer->addChild(_displayValueLabel);
        
        // Add the alert
        Text* alert = Text::create("ScrollView both","fonts/Marker Felt.ttf",30);
        alert->setColor(Color3B(159, 168, 176));
        alert->setPosition(Vec2(widgetSize.width / 2.0f, widgetSize.height / 2.0f - alert->getContentSize().height * 3.075f));
        _uiLayer->addChild(alert);
        
        Layout* root = static_cast<Layout*>(_uiLayer->getChildByTag(81));
        
        Layout* background = static_cast<Layout*>(root->getChildByName("background_Panel"));
        
        // Create the dragpanel
        ui::ScrollView* scrollView = ui::ScrollView::create();
        scrollView->setDirection(ui::ScrollView::Direction::BOTH);
        scrollView->setTouchEnabled(true);
        scrollView->setBounceEnabled(true);
        scrollView->setBackGroundImageScale9Enabled(true);
        scrollView->setBackGroundImage("cocosui/green_edit.png");
        scrollView->setContentSize(Size(210, 122.5));
		scrollView->setScrollBarWidth(4);
		scrollView->setScrollBarPositionFromCorner(Vec2(6, 6));
        Size backgroundSize = background->getContentSize();
        scrollView->setPosition(Vec2((widgetSize.width - backgroundSize.width) / 2.0f +
                                    (backgroundSize.width - scrollView->getContentSize().width) / 2.0f,
                                    (widgetSize.height - backgroundSize.height) / 2.0f +
                                    (backgroundSize.height - scrollView->getContentSize().height) / 2.0f));
        ImageView* imageView = ImageView::create("Hello.png");
        scrollView->addChild(imageView);
        
        scrollView->setInnerContainerSize(imageView->getContentSize());
        Size innerSize = scrollView->getInnerContainerSize();
        imageView->setPosition(Vec2(innerSize.width / 2.0f, innerSize.height / 2.0f));
        
        _uiLayer->addChild(scrollView);

        // Jump to right bottom
        scrollView->jumpToBottomRight();

		//////////////////////////////////////////////////////////////////////////
		// Adding keyboard listener to reproduce scroll stopping issue!
		{
			auto kb_listener = EventListenerKeyboard::create();
			kb_listener->onKeyPressed = [this, scrollView] (EventKeyboard::KeyCode key_code, Event* event)
			{
				if (key_code == EventKeyboard::KeyCode::KEY_ESCAPE)
				{
					scrollView->retain();
					scrollView->removeFromParent();
					_uiLayer->addChild(scrollView);
					scrollView->setTouchEnabled(false);
					scrollView->setTouchEnabled(true);
					scrollView->release();
				}
			};
			Director::getInstance()->getEventDispatcher()->addEventListenerWithSceneGraphPriority(kb_listener, _uiLayer);
		}
		//////////////////////////////////////////////////////////////////////////

        return true;
    }
    
    return false;
}
```

Than run the application, find the required test scene, start moving `ScrollView` and press `ESC` button without releasing the touch. Then release the touch. `ScrollView` will be stuck in scrolling mode.

And here is video demonstrating the issue :)
![untitled](https://user-images.githubusercontent.com/6168693/33843164-49a6ffa2-dea5-11e7-8480-82cff9c6cf57.gif)

**Comment by drelaptop** [comment]
PR: https://github.com/cocos2d/cocos2d-x/pull/18552 is a good patch, and it works well.
Can you enrich the test case? you can add your test code to `bool UIScrollViewTest_Both :: init ()` as explained above. At the same time it is better to add more details of the test case. 

Thanks for your contribution.

**Comment by AIGRIND-LLC** [comment]
Tanks for your feedback @drelaptop 😊

Don't you think the test modification provided above is to confusing for most users?
Especially these two lines that actually are workaround of the internal node implementation:
```
scrollView->setTouchEnabled(false);
scrollView->setTouchEnabled(true);
```
Also it is necessary to design another test UI to reproduce required behaviour because ESC button is not available on mobile devices.

At this moment some tests in **cpp-project** are already designed as "Fix for issue #BLAH-BLAH-BLAH". In most cases theirs implementation is odd and totally obsolete. And you are lucky if you find corresponding issue here at Github. Should we create another one? 🤔

**Comment by drelaptop** [comment]
I agree with you, creating another test is better than using the old. You can create a new test it as the old tests do, "Fix for issue #BLAH-BLAH-BLAH", and add it into tests list  of cpp-test project.

**Comment by AIGRIND-LLC** [comment]
I have created new test and committed it. And it helped me to find one minor bug in my pull request so I have made additional fix. Please check.

![untitled](https://user-images.githubusercontent.com/6168693/33944257-49d11424-e024-11e7-9414-b09b89583a4c.gif)
Also you can find some debug messages in Output log when running the test.

**Comment by drelaptop** [comment]
@AIGRIND-LLC  You are awesome, for the `.gif` images which show the issue and test are perfect, and your PR works well. 

this issue has been solved at #18552, and should be closed @minggo

**Comment by minggo** [comment]
We should close an issue after corresponding PR is merged.
